src/os400qc3.c

```diff
@@ -1003,51 +1003,105 @@ libssh2_os400qc3_hash(const unsigned char *message, unsigned long len,
     return 0;
 }
 
-void
+static int
 libssh2_os400qc3_hmac_init(_libssh2_os400qc3_crypto_ctx *ctx,
                            int algo, size_t minkeylen, void *key, int keylen)
 {
+    Qus_EC_t errcode;
+
     if(keylen < minkeylen) {
         char *lkey = alloca(minkeylen);
 
         /* Pad key with zeroes if too short. */
         if(!lkey)
-            return;
+            return 0;
         memcpy(lkey, (char *) key, keylen);
         memset(lkey + keylen, 0, minkeylen - keylen);
         key = (void *) lkey;
         keylen = minkeylen;
     }
-    libssh2_os400qc3_hash_init(&ctx->hash, algo);
+    else
```

Comment:
This else clause is wrong: the test above is for padding, not for length overflow. The case `keylen >= minkeylen` (32 < 32) then causes my current failure.

Reply:
Oops, pushed a fixed for this!

And for the datalen/len copy-paste mistake.